Fix filedescriptor out of range in select() when running commands

bwoebi opened the pull request:
##### SUMMARY
With a large number of servers involved, we sometimes end up having a pipes with fds >= 1024; i.e. too large for use with select().

This PR proposes to mitigate this issue by using the kqueue/epoll etc. mechanisms if available.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
run_command() from module_utils/basic.py

##### ADDITIONAL INFORMATION
Probably lib/ansible/utils/cmd_functions.py should also be updated. (But it did not affect me right now, so...)

bwoebi committed:
Fix filedescriptor out of range in select() when running commands

bwoebi commented:
There is one empty-init issue ... not sure how to exempt that.

mattclay commented:
A compatibility shim for `lib/ansible/compat/__init__.py` should be kept to maintain backwards compatibility with third party plugins expecting that location.

abadger commented on lib/ansible/compat/__init__.py:
```diff
@@ -1,26 +0,0 @@
-Compat library for ansible.  This contains compatibility definitions for older python
-When we need to import a module differently depending on python version, do it
-here.  Then in the code we can simply import from compat in order to get what we want.
-'''
```
This file should probably not have been deleted/truncated.

abadger replied:
Although, I suppose that we could make it a policy that any backwards compatibility code should go into module_utils so it is usable module-side as well as controller side.

@sivel, @sdoran ^ Do you think it's okay to get rid of the controller side compat and make it policy that all compat code goes into module_utils from now on?

abadger replied:
Ah... I see @mattclay commented that there should be a backwards-compat shim so third party plugins (especially third party connection plugins that may have based their work on the connection plugins we ship) will keep working.  That seems like a good idea and would mean that `compat/__init__.py` needs to remain for now as well.

mattclay replied:
I think generally compat code should go into module_utils rather than being controller-only, as long as it's likely to be usable by modules.

In this case, since we're moving existing compat code we should be sure to keep a shim in the original location for backwards compatibility with existing plugins. We may want a deprecation warning for use of that shim though, so we don't have to maintain it indefinitely.

abadger replied:
PS: The backwards compat shim will look something like the code in ib/ansible/module_utils/compat/selectors/__init__.py

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -2619,19 +2620,20 @@ def run_command(self, args, check_rc=False, close_fds=True, executable=None, dat
                 # the pipes are completely read and
                 # the process is terminated
-                if (not rpipes or not rfds) and cmd.poll() is not None:
+                if len(selector.get_map()) == 0 and cmd.poll() is not None:
```
You should be able to write these conditions as:
``` python
if not selector.get_map() and cmd.poll() is not None:
```

EDIT: properly negated `get_map()`

abadger commented:
Thanks for working on this!  I've wanted to do this for a long time but never got around to it.

I left a few notes.  I see that you inherited the code I'm suggesting be changed from the old read_pipes code but it would be good to clean those up when we switch to selectors (otherwise people will forever wonder if there's a reason that the code is doing those things or if they're unnecessary).

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -2607,9 +2602,15 @@ def run_command(self, args, check_rc=False, close_fds=True, executable=None, dat
-                stderr += self._read_from_pipes(rpipes, rfds, cmd.stderr)
+                events = selector.select(1)
+                for key, event in events:
+                    b_chunk = os.read(key.fileobj.fileno(), self.get_buffer_size(key.fileobj))
```
* Is there a reason for using `os.read()` instead of `key.fileobj.read()`?
* Is there a reason for using `get_buffer_size()` rather than just letting `read()` slurp in all of the data that's available?

bwoebi replied:
I got some failures with tests and I'm not actually that well versed with all the idioms around testing and mocking … thus I just preserved the currently present impl.

abadger replied:
Okay... We shouldn't let tests drive implementation (unless it's a case where the implementation is more modular, easier to read, or more flexible once it's adapted to the test case).  let me take a look at updating the test case.

abadger replied:
I pushed this change in my commit to the code

abadger committed:
Simplify the run_command() code

Now that we're using selectors in run_command(), we can simplify some of
the code.

* Use fileobj.read() instead of os.read()
* No longer use get_buffer_size() as we can just slurp all of the data
  instead.

Also use a simpler conditional check of whether the selector map is
empty

abadger commented:
Okay, I've made my suggested changes to the `run_command` code and updated the tests to work with them.  You'll still need to address mattclay's concern and then this should be in good shape.  

Adding a changelog fragment would also be good in case this is a bugfix that should be backported.

bwoebi commented:
@abadger Thanks for the quick review - Added the BC wrapper and changelog fragment.

Question remains though what I can do about the empty-init CI failure? Where to exempt this?

mattclay commented:
@bwoebi Before dealing with the empty-init sanity test I need to figure out what's going on with the import sanity test.

abadger commented:
@mattclay  I think what's happening is that the `compat.selectors` package tries to hide the fact that it could be implemented by either the python3.5+ selectors package or the selectors2 package.  it does that by importing one of those two modules and then replacing itself in sys.modules with that actual implementation.  (This part is successful).  Then the import sanity test tries to load the `compat.selectors._selectors2` module.  That fails because `compat.selectors` has already replaced itself, making itself a module (single python file) instead of a package (directory containing an `__init__.py`). 

I think the test will pass if we rearrange the files somewhat.  Two options:
1. ```mv lib/ansible/module_utils/compat/selectors/_selectors2.py  lib/ansible/module_utils/compat/selectors/``` And then change the import in `compat/selectors/__init__.py` to point to the new location of _selectors2.py
2. ```mv lib/ansible/module_utils/compat/selectors/*.py lib/ansible/module_utils/compat/``` And then change the import in `compat/selectors.py` to point to the new location of _selectors2.py.

We could also ignore the error for `_selectors2.py` since we know it is bogus in this case (code should only import the selectors name directly, not the _selectors2 name) but that feels like it could let an unrelated problem with imports slip through in the future so it's not an option that I favor.

bwoebi commented:
@abadger So what option is to be preferred?

abadger commented:
I think option (2) is my preference.  @sivel, do you have a preference?

(Note: sivel is on PTO until next week so if you just want to go ahead with Option 2, I'm sure that he'll weigh in if he thinks option 1 is better)

mattclay commented:
I think option 2 is more likely to work, so I'd try that first.

bwoebi commented:
I have no idea what the pylinters issue is here:

```
# test/units/module_utils/basic/test_run_command.py:115:12: bad-super-call: Bad first argument 'MockSelector' given to super()
    class MockSelector(selectors.BaseSelector):
        def __init__(self):
            super(MockSelector, self).__init__() # line 115
```
What should be the first arg, if not the very class name?

mattclay commented on test/units/module_utils/basic/test_run_command.py:
```diff
@@ -41,38 +39,112 @@ def mock_os_abspath(path):
 
-    fake_select = mocker.patch('ansible.module_utils.basic.select')
-    fake_select.select.side_effect = mock_select
+    class MockSelector(selectors.BaseSelector):
```
The pylint error is occurring because this `MockSelector` class shadows the one on line 59.

mattclay commented:
Integration tests are also failing on this PR:

https://app.shippable.com/github/ansible/ansible/runs/156100/88/tests

bwoebi commented:
@mattclay any suggestions on how to debug this? When running the test in isolation it just succeeds. Maybe it's a leftover of some other test, but tried a few - they all were not leaving any residues.

```
$ ansible-test integration remote_tmp                                                               
Running remote_tmp integration test script
+ ansible-playbook -i ../../inventory playbook.yml -e output_dir=/home/dev/private/ansible/test/results/.tmp/output_dir -v
Using /home/dev/private/ansible/test/results/.tmp/integration/remote_tmp-blTReV-ÅÑŚÌβŁÈ/test/integration/integration.cfg as config file
[WARNING]: The `junit_xml` python module is not installed. Disabling the `junit` callback plugin.


PLAY [Test temp dir on de escalation] **************************************************************************************************************************************************************************************************************************************************************************************

TASK [create test user] ****************************************************************************************************************************************************************************************************************************************************************************************************
changed: [testhost] => {"changed": true, "comment": "", "create_home": true, "group": 1005, "home": "/home/tmptest", "name": "tmptest", "shell": "", "state": "present", "stderr": "useradd: warning: the home directory already exists.\nNot copying any file from skel directory into it.\n", "stderr_lines": ["useradd: warning: the home directory already exists.", "Not copying any file from skel directory into it."], "system": false, "uid": 1002}                                                                                                                                                                                            

TASK [Test case from issue 41340] ******************************************************************************************************************************************************************************************************************************************************************************************
changed: [testhost] => {"changed": true, "msg": "File created"}

TASK [clean up file] *******************************************************************************************************************************************************************************************************************************************************************************************************
changed: [testhost] => {"changed": true, "path": "/tmp/testing.txt", "state": "absent"}

TASK [clean up test user] **************************************************************************************************************************************************************************************************************************************************************************************************
changed: [testhost] => {"changed": true, "force": false, "name": "tmptest", "remove": false, "state": "absent"}

PLAY [Test tempdir is removed] *********************************************************************************************************************************************************************************************************************************************************************************************

TASK [file] ****************************************************************************************************************************************************************************************************************************************************************************************************************
changed: [testhost] => {"changed": true, "dest": "//home/dev/private/ansible/test/results/.tmp/output_dir/65393", "gid": 1001, "group": "dev", "mode": "0664", "owner": "dev", "size": 0, "state": "file", "uid": 1001}

TASK [copy] ****************************************************************************************************************************************************************************************************************************************************************************************************************
changed: [testhost] => {"changed": true, "checksum": "da39a3ee5e6b4b0d3255bfef95601890afd80709", "dest": "//home/dev/private/ansible/test/results/.tmp/output_dir/65393.2", "gid": 1001, "group": "dev", "md5sum": "d41d8cd98f00b204e9800998ecf8427e", "mode": "0664", "owner": "dev", "size": 0, "src": "//home/dev/private/ansible/test/results/.tmp/output_dir/65393", "state": "file", "uid": 1001}                                                                                                                                                                                                                                                 

TASK [find] ****************************************************************************************************************************************************************************************************************************************************************************************************************
ok: [testhost] => {"changed": false, "examined": 9, "files": [{"atime": 1579069811.3576586, "ctime": 1579069811.3496587, "dev": 66306, "gid": 1001, "gr_name": "dev", "inode": 16229774, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": false, "isreg": true, "issock": false, "isuid": false, "mode": "0764", "mtime": 1579069811.3456588, "nlink": 1, "path": "/home/dev/.ansible/tmp/ansible-tmp-1579069811.01-262062582285439/AnsiballZ_find.py", "pw_name": "dev", "rgrp": true, "roth": true, "rusr": true, "size": 118058, "uid": 1001, "wgrp": true, "woth": false, "wusr": true, "xgrp": false, "xoth": false, "xusr": true}], "matched": 1, "msg": ""}                                                                                                                                                                                                                                                                         

TASK [debug] ***************************************************************************************************************************************************************************************************************************************************************************************************************
ok: [testhost] => {
    "result": {
        "changed": false, 
        "examined": 9, 
        "failed": false, 
        "files": [
            {
                "atime": 1579069811.3576586, 
                "ctime": 1579069811.3496587, 
                "dev": 66306, 
                "gid": 1001, 
                "gr_name": "dev", 
                "inode": 16229774, 
                "isblk": false, 
                "ischr": false, 
                "isdir": false, 
                "isfifo": false, 
                "isgid": false, 
                "islnk": false, 
                "isreg": true, 
                "issock": false, 
                "isuid": false, 
                "mode": "0764", 
                "mtime": 1579069811.3456588, 
                "nlink": 1, 
                "path": "/home/dev/.ansible/tmp/ansible-tmp-1579069811.01-262062582285439/AnsiballZ_find.py", 
                "pw_name": "dev", 
                "rgrp": true, 
                "roth": true, 
                "rusr": true, 
                "size": 118058, 
                "uid": 1001, 
                "wgrp": true, 
                "woth": false, 
                "wusr": true, 
                "xgrp": false, 
                "xoth": false, 
                "xusr": true
            }
        ], 
        "matched": 1, 
        "msg": ""
    }
}

TASK [assert] **************************************************************************************************************************************************************************************************************************************************************************************************************
ok: [testhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

PLAY RECAP *****************************************************************************************************************************************************************************************************************************************************************************************************************
testhost                   : ok=9    changed=6    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

mattclay commented:
@bwoebi I was able to reproduce the issue with the following command:

`ansible-test integration shippable/posix/group3/ --docker centos6 -v`

So it does appear to be an interaction between at least two different tests. It will probably work as a repro without the `--docker` option as well.

You may want to try isolating the tests which are interacting with each other by using the `--exclude {target}` option to reduce the tests in the group until you've isolated the ones involved.

bwoebi commented:
Thanks, I could isolate it. (run pids followed by remote_tmp)

So, the problem is: https://github.com/ansible/ansible/blob/1db6d5598abb465738a7176d8b7c64a7143c6628/test/integration/targets/pids/tasks/main.yml#L33-L37

async without poll. The task actually terminates safely, but it now waits until the child process has actually exited (which I'd say is the better behavior).
The reason is buried deep in the selectors implementation within signal handling I guess.

The trivial "fix" for the test is to just reduce the async timeout ... (to e.g. 5 seconds)
Because TBH, I'm not sure whether it's actually an issue, for me it's quite logical that the parent command just stays alive as long as the child lives.

Your opinion? @mattclay At least I don't know how I would fix that - selectors is encapsulating that behavior somehow.

bwoebi commented:
bump @mattclay ?

bwoebi commented:
ready_for_review

FliegenKLATSCH commented:
Would be nice if this could be merged. Deploying to large amount of servers is really time consuming and error prone because of this limitation.

abadger commented:
@sivel Would you mind looking this PR over?  I think this is important but since I don't work on this code anymore, it's probably better for someone else to review and merge.  I would ordinarily ping @samdoran about it since it's in module_utils/basic.py but he's on paternity leave so that wouldn't get this looked at in a timely fashion.

samdoran commented:
I'm taking a look at this now.

samdoran reviewed (approved):
These changes look good. I'm re-running tests to make sure they still pass then I'm ok with merging this.

mkrizek commented:
!waffling core_review
